utils/math.go

```diff
@@ -0,0 +1,50 @@
+package utils
+
+import (
+	"fmt"
+
+	"cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// CalculateSharesFromAssets returns the number of shares that correspond
+// to a given amount of deposited assets.
+//
+// If totalAssets is zero (i.e. first deposit), it returns the same amount
+// of shares as assets (1:1 ratio). Otherwise:
+//
+//	shares = (assets * totalShares) / totalAssets
+func CalculateSharesFromAssets(
+	assets math.Int,
+	totalAssets math.Int,
+	totalShares math.Int,
+	shareDenom string,
+) (sdk.Coin, error) {
+	if totalAssets.IsZero() {
+		return sdk.NewCoin(shareDenom, assets), nil // First deposit: 1:1 mapping
```

Comment:
we should verify that we don't need more precision for shares than the underlying denomination... I _suspect_ that 1:1 is fine but we should think through if something like 1:100 might make sense to increase the shares precision ... if we go that way we would need to worry about type size restrictions and overflow though.